### PR TITLE
Demo submission: Automated development pipeline for Chrome extensions

### PR DESCRIPTION
**Pull request by andreaskth**
# Demo submission: Automated development pipeline for Chrome extensions

## Members:

Andreas Kärrby (karrby@kth.se)
Github: [andreaskth](https://github.com/andreaskth)

Henrik Kultala (kultala@kth.se)
Github: [hengque](https://github.com/hengque)

## Proposal/motivation

Original proposal PR is #1076 

## Submission

The video demo can be found at: https://www.youtube.com/watch?v=SKJYKNKyaqI
**EDIT:** New video demo (after feedback) can be found at: https://www.youtube.com/watch?v=sTF_A4SPsh0

## What we did

- Created a Chrome extension that spawns easter eggs on websites, and lets the user collect these eggs ([available here](https://chrome.google.com/webstore/detail/easter-eggstension/lkeplkoegchdobbdodnocghcikijbnfm?hl=en)).
- Created a video demo showcasing an automated CI/CD pipeline for developing extensions for Google Chrome.
  - Work is done in feature branches, and tests are automatically run on commits.
  - Feature branches are merged into the dev-branch, representing an "unofficial" version of the extension, available to trusted testers. Tests are run on PRs to this branch.
  - For official releases (that appear for everyone), the dev-branch is merged into the main-branch.
- Subtitles are available (have to be manually activated by the viewer)
- Sources for all images are disclosed throughout the video, and in the video description.

## Grading criteria we are aiming for

|                                             | Yes | No | Remarkable |
|-------------------------------------------- | ----|----|-------------|
|The demonstration is clearly motivated (why it matters for Devops?) | Yes 🐰 | No | Relates to a hard problem |
|The demonstration is difficult to do | Yes 🐣 | No | Relies on a non trivial infrastructure |
|The demonstration is original | Yes | No | The are less than 10 demos on the topic on Internet 🌷 |
|The video is sublime (eg visually appealing) | Yes | No | Excellent narrative 🐰 |
|The video contains an [easter egg](https://github.com/OrkoHunter/python-easter-eggs) | Yes | No | Related to the demo 🐣 |
|There is a code repo to run the demo  | Yes | No | Code repo with a solid readme |
|The video must contain subtitles which are clear and in proper English | Yes | No | Clearly understandable voice over 🌷 |
|The video includes a take-home message | Yes | No | Actionable takeaway 🐰 |

### CODE REVIEW

**Comment by nwessman** [comment]
# Feedback
A well-made video, with nice graphics and clear speech.  The Chrome extension that you have created looked very appealing, which made it more fun to follow along with the demo. Like most 3-minutes demo, I wish the demo could have been longer since most of the problems with the demo has to do with it speeds through aspects that I would have wanted to see more in-depth, but that is of course not your fault.
 
I think it was very good that you included small tips like that Google will not allow you to publish new versions if the version number has not been increased. This also made your Github Actions seem very relevant. 

Since I have never developed a Chrome extension, I was a bit unsure which version number had to be changed and where to do that, since you only showed the Github Action-file for checking that the number had changed. That might be out of the scope of the video and the video highlighted the problem so I can look up that information myself. This maybe is a trivial problem and self-explanatory for an extension developer but it sounded very important to the automated process so I had to double-check that I did not miss anything.

Sometimes you say something but show something else, in the beginning, you say that you use Github Workflows but you show the Github Actions logo. I know that workflows are a part of Actions, but it makes it a bit less clear.

Sine if the websites you are visiting are a bit cluttered with information. For example the website at 1:36 there is a whole column on the right that is full of information that is not part of the demo. Here and in similar parts of the demo you could zoom in on the relevant part of the sites so you do not lose the attention of the watcher on other irrelevant information.

Overall, it was a very nice video, which demonstrated the process in an easily accessible manner. It has convinced me to set up the same automated process if I would start developing for the Chrome Store.

Good job! :trophy:

**Comment by elmaxe** [comment]
Good job!

A very graphically pleasing video, with a nice voice over! The Chrome extension was humorous 👍🏻 
Sometimes you show too much of the screen. Do you really need to always show the whole screen? Sometimes you should display only what you are focusing on.
In the end of the video you summarize. This part is clearly sped up more than the rest of the video. Perhaps you should speed up the whole video a little bit instead of speeding up the last 9 seconds a lot. It would be nice if the video could be longer and more thorough, but we all know that isn't allowed, so I don't blame you for that.

Great job! 😍

**Comment by oscaralmqvist** [comment]
Good job! I felt that it was good introduction which gives an overview of the proposal. There's also a voiceover along with the subtitles, which is always appreciated. The animations you used are really good and makes it easy to follow the tree structure of a repository. It was a bit hard to follow the project creation on GCP and also the purpose of, for example, enabling the Web Store API. Another point I feel could be improved are the explanations of the GitHub Actions. I felt that they could be a bit more indepth about what the different steps are. All in all, I enjoyed this tutorial! 🎉

**Comment by oskstr** [comment]
## Feedback

Very nice video. Interesting topic, clear voice over, subtitles, and for the intro - nice animations. 

Something that could be better is probably the part with setting up tokens and users, it felt a bit rushed with text flashing by. I understand that it was probably because of the restricted time of the video demo. If anything, you could perhaps cut out the obvious parts (like entering contact information) to give us more time looking at the 

Also, the voice over was at times disconnected from what was happening at the screen. Like "Here we see the consent screen we configured earlier." and after that the consent screen is actually shown. 

But overall it was good. I have thought about setting up Chrome Extensions (Eggstenstions 😂) myself before but never really looked into how it is actually done.  🚀

**Comment by damnePers** [comment]
Thanks for the demo, here are some feedback:

First of all, your demo is too long. Time 3:22, shorten this to get to 3 minutes since that was a strict limit. 

It is good that you start with what the demo is about and what you have done as prerequisites. Also, it is nice that you show and define your pipeline which gives the viewer a good overview of what's to come. However, this is your main focus, it shouldn’t be. Your video feels more like a presentation about the pipeline than an actual demo of how to set it up and use it. I would suggest that you add, in the first part - up until 2:17, more actionable steps. Instead of just showing code, show where to add it, how to write it/get it etc. Show how to 

Regarding the beginning part where you have slides with a lot of text as you speak, it is really difficult to follow along in the video as well as reading the slides. Here is a good part where you could switch out the slides for actionable steps instead. 

Emphasis what part about your demo actually integrates multi.account. As I see it you only show how to set up a pipeline and not how to connect multiple accounts. 

It would be nice if there were some documentation where you list what you are doing in the demo so that the viewer could easily access the commands and code used in the demo. There you could also list what prerequisites are needed for this demo and instructions for the steps you make in your demo so the viewer could look up any details they might have missed or find difficult. 

Also, link to the example repo you mention in your proposal.

**Comment by bubriks** [comment]
Nice work on the demo! I like the unique topic, however, the YAML files and the second part of the video seem a bit rushed. This of course is due to the limited time so there is not that much that can be done about it.

**Comment by andreaskth** [comment]
Thanks a lot for the feedback everyone! We have gone through your comments and implemented the following changes:

* Removed the part explaining how extensions are made (would be nice to include, but given the time limit we preferred to remove it so that the rest of the video can be less rushed)
* Re-recorded the audio and tweaked a few words so that it's a more relaxed pace and also so that the voice-over matches what is shown in the video (for example "here we see the consent screen configured earlier" is actually said when the consent screen is shown, and so on).
* Made a list with bullet points at the end of the video to emphasize the key take-aways from the video
* Added some pleasant background music to make the video more enjoyable to watch
* Faded irrelevant parts of the screen black, to highlight what is important/what the viewer should look at specifically.

The new video is available here: https://www.youtube.com/watch?v=sTF_A4SPsh0
And new PR is here: https://github.com/KTH/devops-course/pull/1520
